presto-cli/src/main/java/io/prestosql/cli/OutputHandler.java

```diff
@@ -13,85 +13,79 @@
 import static java.util.Objects.requireNonNull;
-import static java.util.concurrent.TimeUnit.SECONDS;
 
 public final class OutputHandler
         implements Closeable
 {
-    private static final Duration MAX_BUFFER_TIME = new Duration(3, SECONDS);
-    private static final int MAX_BUFFERED_ROWS = 10_000;
+    private static final int MAX_QUEUED_ROWS = 50_000;
 
     private final AtomicBoolean closed = new AtomicBoolean();
-    private final List<List<?>> rowBuffer = new ArrayList<>(MAX_BUFFERED_ROWS);
+    private final BlockingQueue<List<?>> rowQueue = new LinkedBlockingQueue<>(MAX_QUEUED_ROWS);
     private final OutputPrinter printer;
 
-    private long bufferStart;
+    private CompletableFuture<Void> future;
 
     public OutputHandler(OutputPrinter printer)
     {
         this.printer = requireNonNull(printer, "printer is null");
     }
 
-    public void processRow(List<?> row)
-            throws IOException
-    {
-        if (rowBuffer.isEmpty()) {
-            bufferStart = System.nanoTime();
-        }
-
-        rowBuffer.add(row);
-        if (rowBuffer.size() >= MAX_BUFFERED_ROWS) {
-            flush(false);
-        }
-    }
-
     @Override
     public void close()
             throws IOException
     {
         if (!closed.getAndSet(true)) {
-            flush(true);
             printer.finish();
         }
     }
 
     public void processRows(StatementClient client)
             throws IOException
     {
-        while (client.isRunning()) {
-            Iterable<List<Object>> data = client.currentData().getData();
-            if (data != null) {
-                for (List<Object> row : data) {
-                    processRow(unmodifiableList(row));
+        this.future = CompletableFuture.runAsync(() -> {
+            while (client.isRunning()) {
+                Iterable<List<Object>> data = client.currentData().getData();
+                if (data != null) {
+                    for (List<Object> row : data) {
+                        try {
+                            rowQueue.put(unmodifiableList(row));
+                        }
+                        catch (Exception ex) {
+                            throw new RuntimeException(ex);
+                        }
+                    }
                 }
+                client.advance();
             }
-
-            if (nanosSince(bufferStart).compareTo(MAX_BUFFER_TIME) >= 0) {
-                flush(false);
+        });
+        while (!future.isDone()) {
+            while (!rowQueue.isEmpty()) {
+                printer.printRow(rowQueue.poll(), false);
             }
-
-            client.advance();
         }
-    }
-
-    private void flush(boolean complete)
-            throws IOException
-    {
-        if (!rowBuffer.isEmpty()) {
-            printer.printRows(unmodifiableList(rowBuffer), complete);
-            rowBuffer.clear();
+        while (!rowQueue.isEmpty()) {
+            printer.printRow(rowQueue.poll(), false);
```

Comment:
Rather than removing and printing one row at a time, we can fetch a large batch at once. The `Queues.drain()` method from Guava will handle this (by calling `drainTo()` internally) along with a timeout, that we can use to preserve the `MAX_BUFFER_TIME` behavior as explained above.

After changing to this model, we won't need the new `printRow()` method.